File: longest_sub_string.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
int length(char * s);
int main()
{
    char word[] = "abcabcd";
    printf("%d",length(word));
    return 0;
}
int length(char * s)
{
    int lastindex[256];
    int start=0;
    int max=0;
    memset(lastindex,-1,sizeof(lastindex));
    for(int i=0;s[i]!='\0';i++)
    {
        unsigned char c = (unsigned char)s[i] ;
        if (lastindex[c] >= start)
        {
            start = lastindex[c] + 1;
        }
        int cur = i - start + 1;
        if(cur > max)
        {
            max=cur;
        }
        lastindex[c] = i;
        
    }
    
    return max;
}

File: longest_substring.py
word = "abcdefabcdefghi"
def length(s):
    start = 0
    cur = 0
    max = 0
    lastindex = []
    sub = []
    maxstart=0
    for i in range(0, 256):
        lastindex.append(-1)
    for i in range(0, len(s)):
        c = ord(s[i])
        if lastindex[c] >= start:
            start = lastindex[c] + 1
        cur = i - start + 1
        if cur > max:
            max = cur
            maxstart=start
        lastindex[c] = i
    print(maxstart)
    for i in range(maxstart, maxstart+max):
        sub.append(s[i])
    print(sub)
    return max

print(length(word))
